doc/python: add section about replaced/stubbed/avoided packages (#336775)
NixOS/nixpkgs: doc/languages-frameworks/python.section.md

  and using a `-` as delimiter.
* Attribute names in `python-packages.nix` should be sorted alphanumerically to
  avoid merge conflicts and ease locating attributes.
* The list of [dependencies to avoid or replace](#python-dependencies-avoid-replace) is respected.

This list is useful for reviewers as well as for self-checking when submitting packages.

> ## Dependencies to avoid or replace {#python-dependencies-avoid-replace}

Reviewer: To scope it within nixpkgs contributions, rather than for example downstream usage like ad-hoc shells
```suggestion
### Dependencies to avoid or replace {#python-dependencies-avoid-replace}
```

Some dependencies do not make sense inside a nix build and should be completely avoided or replaced by a stub.
Below you find a listing of the most common ones.

| Package | Replacement | Reason |
|---------|-------------|--------|
| cmake | cmake (stub) | The package normally downloads cmake from the internet. The stub provides our cmake with the usual setup hook. |

> | ninja | ninja (stub) | The package normally downloads ninja from the internet. The stub provides our cmake with the usual setup hook. |

Reviewer 2: This seems more like an implementation detail than actionable advice for package authors, right?

Author: Well, jain. It is probably a good hint to kinda know the difference between pkgs.ninja and pkgs.python3Packages.ninja and what to do when you encounter it.

Author: For example with this information people know that they don't need to remove the dependency which would otherwise be necessary because downloading things in the sandbox is well known to not work.

> | cmake | cmake (stub) | The package normally downloads cmake from the internet. The stub provides our cmake with the usual setup hook. |
> | ninja | ninja (stub) | The package normally downloads ninja from the internet. The stub provides our ninja with the usual setup hook. |
> | pytest-cov | pytest-cov-stub | coverage is only useful for developers. |
> | pytest-runner | | Dependencies cannot be downloaded and should be provided through nativeCheckInputs. |

Reviewer: ```suggestion
| cmake | `python3Packages.cmake` (stub) | The package normally downloads cmake from the internet. The stub provides our cmake with the usual setup hook. |
| ninja | `python3Packages.ninja` (stub) | The package normally downloads ninja from the internet. The stub provides our ninja with the usual setup hook. |
| pytest-cov | `python3Packages.pytest-cov-stub` | coverage is only useful for developers. |
| pytest-runner | - | Dependencies cannot be downloaded and should be provided through nativeCheckInputs. |
```

* Non-python runtime dependencies should be added via explicit wrapping or
  patching (using e.g. `substituteInPlace`), rather than through propagation via
  `dependencies`/`propagatedBuildInputs`, to reduce clutter in `$PATH`.
* The list of [dependencies to avoid or replace](#python-dependencies-avoid-replace) is respected.

This list is useful for reviewers as well as for self-checking when submitting packages.

### Dependencies to avoid or replace {#python-dependencies-avoid-replace}

Some dependencies do not make sense inside a nix build and should be completely avoided or replaced by a stub.
Below you find a listing of the most common ones.

| Package | Replacement | Reason |
|---------|-------------|--------|
| cmake | `python3Packages.cmake` (stub) | The package normally downloads cmake from the internet. The stub provides our cmake with the usual setup hook. |
| ninja | `python3Packages.ninja` (stub) | The package normally downloads ninja from the internet. The stub provides our ninja with the usual setup hook. |

> | pytest-cov | `python3Packages.pytest-cov-stub` | coverage is only useful for developers. |

Reviewer: ```suggestion
| pytest-cov | `python3Packages.pytest-cov-stub` | coverage is mainly useful for development not distribution. |
```
